### Installation

In [1]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth
else:
    # Do this only in Colab notebooks! Otherwise use pip install unsloth
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft trl==0.15.2 triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf datasets huggingface_hub hf_transfer
    !pip install --no-deps unsloth

### Unsloth

In [2]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 2048 # Choose any! We auto support RoPE Scaling internally!
dtype = None # None for auto detection. Float16 for Tesla T4, V100, Bfloat16 for Ampere+
load_in_4bit = True # Use 4bit quantization to reduce memory usage. Can be False.

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/Meta-Llama-3.1-8B-bnb-4bit",      # Llama-3.1 15 trillion tokens model 2x faster!
    "unsloth/Meta-Llama-3.1-8B-Instruct-bnb-4bit",
    "unsloth/Meta-Llama-3.1-70B-bnb-4bit",
    "unsloth/Meta-Llama-3.1-405B-bnb-4bit",    # We also uploaded 4bit for 405b!
    "unsloth/Mistral-Nemo-Base-2407-bnb-4bit", # New Mistral 12b 2x faster!
    "unsloth/Mistral-Nemo-Instruct-2407-bnb-4bit",
    "unsloth/mistral-7b-v0.3-bnb-4bit",        # Mistral v3 2x faster!
    "unsloth/mistral-7b-instruct-v0.3-bnb-4bit",
    "unsloth/Phi-3.5-mini-instruct",           # Phi-3.5 2x faster!
    "unsloth/Phi-3-medium-4k-instruct",
    "unsloth/gemma-2-9b-bnb-4bit",
    "unsloth/gemma-2-27b-bnb-4bit",            # Gemma 2x faster!
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/Meta-Llama-3.1-8B",
    max_seq_length = max_seq_length,
    dtype = dtype,
    load_in_4bit = load_in_4bit,
    # token = "hf_...", # use one if using gated models like meta-llama/Llama-2-7b-hf
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
Unsloth: Failed to patch Gemma3ForConditionalGeneration.
🦥 Unsloth Zoo will now patch everything to make training faster!
==((====))==  Unsloth 2025.3.19: Fast Llama patching. Transformers: 4.51.1.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


model.safetensors:   0%|          | 0.00/5.96G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/235 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/50.6k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/459 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.2M [00:00<?, ?B/s]

We now add LoRA adapters so we only need to update 1 to 10% of all parameters!

In [3]:
model = FastLanguageModel.get_peft_model(
    model,
    r = 16, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = ["q_proj", "k_proj", "v_proj", "o_proj",
                      "gate_proj", "up_proj", "down_proj",],
    lora_alpha = 16,
    lora_dropout = 0, # Supports any, but = 0 is optimized
    bias = "none",    # Supports any, but = "none" is optimized
    # [NEW] "unsloth" uses 30% less VRAM, fits 2x larger batch sizes!
    use_gradient_checkpointing = "unsloth", # True or "unsloth" for very long context
    random_state = 3407,
    use_rslora = False,  # We support rank stabilized LoRA
    loftq_config = None, # And LoftQ
)

Unsloth 2025.3.19 patched 32 layers with 32 QKV layers, 32 O layers and 32 MLP layers.


<a name="Data"></a>
### Data Prep
We now use the Alpaca dataset from [yahma](https://huggingface.co/datasets/yahma/alpaca-cleaned), which is a filtered version of 52K of the original [Alpaca dataset](https://crfm.stanford.edu/2023/03/13/alpaca.html). You can replace this code section with your own data prep.

**[NOTE]** To train only on completions (ignoring the user's input) read TRL's docs [here](https://huggingface.co/docs/trl/sft_trainer#train-on-completions-only).

**[NOTE]** Remember to add the **EOS_TOKEN** to the tokenized output!! Otherwise you'll get infinite generations!

If you want to use the `llama-3` template for ShareGPT datasets, try our conversational [notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Alpaca.ipynb)

For text completions like novel writing, try this [notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Mistral_(7B)-Text_Completion.ipynb).

In [4]:
alpaca_prompt = """Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
{}

### Input:
{}

### Response:
{}"""

EOS_TOKEN = tokenizer.eos_token # Must add EOS_TOKEN
def formatting_prompts_func(examples):
    texts = []
    # Iterate directly through the relevant columns from the CSV
    for target, input_text, stance in zip(examples["target"], examples["text"], examples["stance"]):
        # Create the instruction dynamically
        instruction = "Analyze the stance and identify the target in the following text."
        # Combine target and stance for the response
        output_combined = f"Target: {target}, Stance: {stance}"

        # Must add EOS_TOKEN, otherwise your generation will go on forever!
        text = alpaca_prompt.format(instruction, input_text, output_combined) + EOS_TOKEN
        texts.append(text)
    return { "text" : texts, }

from datasets import load_dataset
# Load the local CSV dataset
dataset = load_dataset("csv", data_files="train_dataset.csv", split="train")
dataset = dataset.map(formatting_prompts_func, batched = True,)


Generating train split: 0 examples [00:00, ? examples/s]

Map:   0%|          | 0/6480 [00:00<?, ? examples/s]

<a name="Train"></a>
### Train the model
Now let's use Huggingface TRL's `SFTTrainer`! More docs here: [TRL SFT docs](https://huggingface.co/docs/trl/sft_trainer). We do 60 steps to speed things up, but you can set `num_train_epochs=1` for a full run, and turn off `max_steps=None`. We also support TRL's `DPOTrainer`!

In [5]:
from trl import SFTTrainer
from transformers import TrainingArguments
from unsloth import is_bfloat16_supported

trainer = SFTTrainer(
    model = model,
    tokenizer = tokenizer,
    train_dataset = dataset,
    dataset_text_field = "text",
    max_seq_length = max_seq_length,
    dataset_num_proc = 2,
    packing = False, # Can make training 5x faster for short sequences.
    args = TrainingArguments(
        per_device_train_batch_size = 2,
        gradient_accumulation_steps = 4,
        warmup_steps = 5,
        # num_train_epochs = 1, # Set this for 1 full training run.
        max_steps = 60,
        learning_rate = 2e-4,
        fp16 = not is_bfloat16_supported(),
        bf16 = is_bfloat16_supported(),
        logging_steps = 1,
        optim = "adamw_8bit",
        weight_decay = 0.01,
        lr_scheduler_type = "linear",
        seed = 3407,
        output_dir = "outputs",
        report_to = "none", # Use this for WandB etc
    ),
)

Unsloth: Tokenizing ["text"] (num_proc=2):   0%|          | 0/6480 [00:00<?, ? examples/s]

In [6]:
# @title Show current memory stats
gpu_stats = torch.cuda.get_device_properties(0)
start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
print(f"GPU = {gpu_stats.name}. Max memory = {max_memory} GB.")
print(f"{start_gpu_memory} GB of memory reserved.")

GPU = Tesla T4. Max memory = 14.741 GB.
7.135 GB of memory reserved.


In [7]:
trainer_stats = trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 6,480 | Num Epochs = 1 | Total steps = 60
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 4 x 1) = 8
 "-____-"     Trainable parameters = 41,943,040/8,000,000,000 (0.52% trained)


Unsloth: Will smartly offload gradients to save VRAM!


In [9]:
# @title Show final memory and time stats
used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
used_percentage = round(used_memory / max_memory * 100, 3)
lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
print(f"{trainer_stats.metrics['train_runtime']} seconds used for training.")
print(
    f"{round(trainer_stats.metrics['train_runtime']/60, 2)} minutes used for training."
)
print(f"Peak reserved memory = {used_memory} GB.")
print(f"Peak reserved memory for training = {used_memory_for_lora} GB.")
print(f"Peak reserved memory % of max memory = {used_percentage} %.")
print(f"Peak reserved memory for training % of max memory = {lora_percentage} %.")

377.9346 seconds used for training.
6.3 minutes used for training.
Peak reserved memory = 7.135 GB.
Peak reserved memory for training = 0.0 GB.
Peak reserved memory % of max memory = 48.402 %.
Peak reserved memory for training % of max memory = 0.0 %.


<a name="Inference"></a>
### Inference
Let's run the model! You can change the instruction and input - leave the output blank!

**[NEW] Try 2x faster inference in a free Colab for Llama-3.1 8b Instruct [here](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Unsloth_Studio.ipynb)**

In [13]:
import pandas as pd
import re

# Load test dataset
test_dataset = pd.read_csv("test_dataset.csv")

# Function to parse model output to extract target and stance
def parse_model_output(output):
    try:
        # Expected format: "Target: {target}, Stance: {stance}"
        target_match = re.search(r"Target:\s*(.*?),\s*Stance:", output)
        stance_match = re.search(r"Stance:\s*(.*?)(?:\s*$|\n)", output)

        target = target_match.group(1).strip() if target_match else "Not Found"
        stance = stance_match.group(1).strip() if stance_match else "Not Found"

        return target, stance
    except Exception as e:
        print(f"Error parsing output: {output}")
        print(f"Error: {e}")
        return "Error", "Error"

# Prepare results storage
results = []

# Enable inference mode
FastLanguageModel.for_inference(model)

# Process each test example
for idx, row in test_dataset.iterrows():
    input_text = row['text']

    # Format input using the same template as training
    prompt = alpaca_prompt.format(
        "Analyze the stance and identify the target in the following text.",  # instruction
        input_text,  # input
        ""  # output - leave blank for generation
    )

    # Tokenize and generate
    inputs = tokenizer([prompt], return_tensors="pt").to("cuda")
    outputs = model.generate(
        **inputs,
        max_new_tokens=128,
        pad_token_id=tokenizer.eos_token_id,
        use_cache=True
    )
    # Decode output
    generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

    # Extract the generated response part (after "### Response:")
    response_part = generated_text.split("### Response:")[-1].strip()

    # Parse the response to get predicted target and stance
    pred_target, pred_stance = parse_model_output(response_part)

    # Store results
    results.append({
        "Original Text": input_text,
        "Predicted Target": pred_target,
        "Predicted Stance": pred_stance,
        "Ground Truth Target": row['target'],
        "Ground Truth Stance": row['stance']
    })

    # Print progress
    if (idx + 1) % 10 == 0:
        print(f"Processed {idx + 1}/{len(test_dataset)} examples")
# Convert results to DataFrame and save to Excel
results_df = pd.DataFrame(results)
results_df.to_excel("test_results.xlsx", index=False)
print("\nResults saved to test_results.xlsx")

# Print some statistics
total = len(results_df)
target_matches = (results_df["Predicted Target"] == results_df["Ground Truth Target"]).sum()
stance_matches = (results_df["Predicted Stance"] == results_df["Ground Truth Stance"]).sum()
print(f"\nAccuracy Statistics:")
print(f"Target Accuracy: {target_matches/total*100:.2f}%")
print(f"Stance Accuracy: {stance_matches/total*100:.2f}%")

Processed 10/1620 examples
Processed 20/1620 examples
Processed 30/1620 examples
Processed 40/1620 examples
Processed 50/1620 examples
Processed 60/1620 examples
Processed 70/1620 examples
Processed 80/1620 examples
Processed 90/1620 examples
Processed 100/1620 examples
Processed 110/1620 examples
Processed 120/1620 examples
Processed 130/1620 examples
Processed 140/1620 examples
Processed 150/1620 examples
Processed 160/1620 examples
Processed 170/1620 examples
Processed 180/1620 examples
Processed 190/1620 examples
Processed 200/1620 examples
Processed 210/1620 examples
Processed 220/1620 examples
Processed 230/1620 examples
Processed 240/1620 examples
Processed 250/1620 examples
Processed 260/1620 examples
Processed 270/1620 examples
Processed 280/1620 examples
Processed 290/1620 examples
Processed 300/1620 examples
Processed 310/1620 examples
Processed 320/1620 examples
Processed 330/1620 examples
Processed 340/1620 examples
Processed 350/1620 examples
Processed 360/1620 examples
P

: You can also use a `TextStreamer` for continuous inference - so you can see the generation token by token, instead of waiting the whole time!

In [16]:

import pandas as pd
import re

def evaluate_model(model, tokenizer, test_dataset, model_name=""):
    """
    Evaluate a model on the test dataset and return results
    """
    # Function to parse model output to extract target and stance
    def parse_model_output(output):
        try:
            # Expected format: "Target: {target}, Stance: {stance}"
            target_match = re.search(r"Target:\s*(.*?),\s*Stance:", output)
            stance_match = re.search(r"Stance:\s*(.*?)(?:\s*$|\n)", output)

            target = target_match.group(1).strip() if target_match else "Not Found"
            stance = stance_match.group(1).strip() if stance_match else "Not Found"

            return target, stance
        except Exception as e:
            print(f"Error parsing output: {output}")
            print(f"Error: {e}")
            return "Not Found", "Not Found"

    # Prepare results storage
    results = []

    # Enable inference mode
    FastLanguageModel.for_inference(model)

    print(f"\nEvaluating {model_name}...")

    # System prompt to guide the model's response format
    system_prompt = """You are a stance detection model. For the given text, you must:
1. Identify the target of the stance
2. Determine the stance (FAVOR, AGAINST, or NONE)
3. Format your response exactly as: "Target: <target>, Stance: <stance>"
For example:
Input: "I love using electric cars, they're great for the environment"
Response: "Target: electric cars, Stance: FAVOR"
"""

    # Process each test example
    for idx, row in test_dataset.iterrows():
        input_text = row['text']

        # Format input using the same template as training but include system prompt for base model
        if model_name == "Base Model":
            prompt = f"{system_prompt}\n\n" + alpaca_prompt.format(
                "Analyze the stance and identify the target in the following text.",
                input_text,
                ""
            )
        else:
            prompt = alpaca_prompt.format(
                "Analyze the stance and identify the target in the following text.",
                input_text,
                ""
            )

        try:
            # Tokenize and generate
            inputs = tokenizer([prompt], return_tensors="pt").to("cuda")
            outputs = model.generate(
                **inputs,
                max_new_tokens=128,
                pad_token_id=tokenizer.eos_token_id,
                use_cache=True,
                temperature=0.1  # Lower temperature for more consistent outputs
            )

            # Decode output
            generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

            # Extract the generated response part (after "### Response:")
            response_parts = generated_text.split("### Response:")
            if len(response_parts) > 1:
                response_part = response_parts[-1].strip()
            else:
                response_part = generated_text.strip()

            # Parse the response to get predicted target and stance
            pred_target, pred_stance = parse_model_output(response_part)

        except Exception as e:
            print(f"Error processing example {idx}: {e}")
            pred_target, pred_stance = "Error", "Error"

        # Store results
        results.append({
            "Original Text": input_text,
            "Predicted Target": pred_target,
            "Predicted Stance": pred_stance,
            "Ground Truth Target": row['target'],
            "Ground Truth Stance": row['stance']
        })

        # Print progress and save intermediate results every 100 rows
        if (idx + 1) % 10 == 0:
            print(f"Processed {idx + 1}/{len(test_dataset)} examples")

        if (idx + 1) % 100 == 0:
            # Save intermediate results
            interim_df = pd.DataFrame(results)
            interim_filename = f"test_results_{model_name.lower().replace(' ', '_')}_rows_1_to_{idx+1}.xlsx"
            interim_df.to_excel(interim_filename, index=False)
            print(f"\nIntermediate results saved to {interim_filename}")

            # Calculate and show interim statistics
            total = len(interim_df)
            target_matches = (interim_df["Predicted Target"] == interim_df["Ground Truth Target"]).sum()
            stance_matches = (interim_df["Predicted Stance"] == interim_df["Ground Truth Stance"]).sum()

            print(f"\nInterim Statistics (First {idx+1} examples):")
            print(f"Target Accuracy: {(target_matches/total*100):.2f}%")
            print(f"Stance Accuracy: {(stance_matches/total*100):.2f}%")

    # Convert final results to DataFrame
    results_df = pd.DataFrame(results)

    # Calculate final statistics
    total = len(results_df)
    target_matches = (results_df["Predicted Target"] == results_df["Ground Truth Target"]).sum()
    stance_matches = (results_df["Predicted Stance"] == results_df["Ground Truth Stance"]).sum()

    target_accuracy = target_matches/total*100
    stance_accuracy = stance_matches/total*100

    print(f"\n{model_name} Final Accuracy Statistics:")
    print(f"Target Accuracy: {target_accuracy:.2f}%")
    print(f"Stance Accuracy: {stance_accuracy:.2f}%")

    return results_df, target_accuracy, stance_accuracy

# Load test dataset
test_dataset = pd.read_csv("test_dataset.csv")

# First, evaluate the base model
print("\nLoading base model...")
base_model, base_tokenizer = FastLanguageModel.from_pretrained(
    model_name="unsloth/Meta-Llama-3.1-8B",
    max_seq_length=max_seq_length,
    dtype=dtype,
    load_in_4bit=load_in_4bit,
)

base_results, base_target_acc, base_stance_acc = evaluate_model(
    base_model,
    base_tokenizer,
    test_dataset,
    "Base Model"
)

# Save base model results
base_results.to_excel("test_results_base_model.xlsx", index=False)
print("\nBase model results saved to test_results_base_model.xlsx")



Loading base model...
==((====))==  Unsloth 2025.3.19: Fast Llama patching. Transformers: 4.51.1.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!

Evaluating Base Model...
Processed 10/1620 examples
Processed 20/1620 examples
Processed 30/1620 examples
Processed 40/1620 examples
Processed 50/1620 examples
Processed 60/1620 examples
Processed 70/1620 examples
Processed 80/1620 examples
Processed 90/1620 examples
Processed 100/1620 examples

Intermediate results saved to test_results_base_model_rows_1_to_100.xlsx

Interim Statistics (First 100 examples):
Target Accuracy: 1.00%
Stance Accuracy: 66.00%
Processed 110/1620 examples
Processed 120/1620 examples
Proc

In [11]:
# alpaca_prompt = Copied from above
FastLanguageModel.for_inference(model) # Enable native 2x faster inference
inputs = tokenizer(
[
    alpaca_prompt.format(
        "Analyze the stance and identify the target in the following text.", # instruction
        "trump might win this elections", # input
        "", # output - leave this blank for generation!
    )
], return_tensors = "pt").to("cuda")

from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer)
_ = model.generate(**inputs, streamer = text_streamer, max_new_tokens = 128)

<|begin_of_text|>Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
Analyze the stance and identify the target in the following text.

### Input:
trump might win this elections

### Response:
Target: election, Stance: FAVOR<|end_of_text|>


<a name="Save"></a>
### Saving, loading finetuned models
To save the final model as LoRA adapters, either use Huggingface's `push_to_hub` for an online save or `save_pretrained` for a local save.

**[NOTE]** This ONLY saves the LoRA adapters, and not the full model. To save to 16bit or GGUF, scroll down!

In [14]:
model.save_pretrained("lora_model")  # Local saving
tokenizer.save_pretrained("lora_model")
# model.push_to_hub("your_name/lora_model", token = "...") # Online saving
# tokenizer.push_to_hub("your_name/lora_model", token = "...") # Online saving

('lora_model/tokenizer_config.json',
 'lora_model/special_tokens_map.json',
 'lora_model/tokenizer.json')

Now if you want to load the LoRA adapters we just saved for inference, set `False` to `True`:

In [15]:
if False:
    from unsloth import FastLanguageModel
    model, tokenizer = FastLanguageModel.from_pretrained(
        model_name = "lora_model", # YOUR MODEL YOU USED FOR TRAINING
        max_seq_length = max_seq_length,
        dtype = dtype,
        load_in_4bit = load_in_4bit,
    )
    FastLanguageModel.for_inference(model) # Enable native 2x faster inference

# alpaca_prompt = You MUST copy from above!

inputs = tokenizer(
[
    alpaca_prompt.format(
        "Analyze the stance and identify the target in the following text.", # instruction
        "trump might win this elections", # input
        "", # output - leave this blank for generation!
    )
], return_tensors = "pt").to("cuda")

from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer)
_ = model.generate(**inputs, streamer = text_streamer, max_new_tokens = 128)

<|begin_of_text|>Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
Analyze the stance and identify the target in the following text.

### Input:
trump might win this elections

### Response:
Target: trump, Stance: FAVOR<|end_of_text|>


You can also use Hugging Face's `AutoModelForPeftCausalLM`. Only use this if you do not have `unsloth` installed. It can be hopelessly slow, since `4bit` model downloading is not supported, and Unsloth's **inference is 2x faster**.

In [ ]:
if False:
    # I highly do NOT suggest - use Unsloth if possible
    from peft import AutoPeftModelForCausalLM
    from transformers import AutoTokenizer
    model = AutoPeftModelForCausalLM.from_pretrained(
        "lora_model", # YOUR MODEL YOU USED FOR TRAINING
        load_in_4bit = load_in_4bit,
    )
    tokenizer = AutoTokenizer.from_pretrained("lora_model")

### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16 or `merged_4bit` for int4. We also allow `lora` adapters as a fallback. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [ ]:
# Merge to 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_16bit", token = "")

# Merge to 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# Just LoRA adapters
if False: model.save_pretrained_merged("model", tokenizer, save_method = "lora",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "lora", token = "")

### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now! We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

[**NEW**] To finetune and auto export to Ollama, try our [Ollama notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)

In [ ]:
# Save to 8bit Q8_0
if False: model.save_pretrained_gguf("model", tokenizer,)
# Remember to go to https://huggingface.co/settings/tokens for a token!
# And change hf to your username!
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "f16", token = "")

# Save to q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "q4_k_m", token = "")

# Save to multiple GGUF options - much faster if you want multiple!
if False:
    model.push_to_hub_gguf(
        "hf/model", # Change hf to your username!
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",],
        token = "",
    )

Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
